File: 112-path-sum/112-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    
    bool fun(TreeNode* root,int sum)
    {
//         if(root==NULL)
//         {
//             if(sum < 0)
//             {
//                 return false;
//             }
            
//             if(sum==0)
//             {
//                 return true;
//             }
//             return false;
//         }
        
        
        
        if(root->left==NULL && root->right==NULL)
        {
            if(sum==0)
            {
                return true;
            }
            return false;
        }
    
       
        
        bool a,b;
        if(root->left)
        {
             a= fun(root->left,(sum-root->left->val));  
        }
         
        if(root->right)
        {
              b= fun(root->right,(sum-root->right->val));
        }
       
        return a | b ; 
        
        
        
    }
 
    bool hasPathSum(TreeNode* root, int targetSum) {
        if(root==NULL)
        {
            return false;
        }
        return fun(root,targetSum-root->val);
    }
};

File: 112-path-sum/112-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    boolean fun(TreeNode root,int tsum)
    {
        if(root==null)
        {
            return false;
        }
        
        if(root.left==null && root.right==null)
        {
            if(tsum==root.val)
            {
                return true;
            }
            else
            {
                return false;
            }
        }
       boolean a= fun(root.left,tsum-root.val);
       boolean b= fun(root.right,tsum-root.val);
        return a | b;
        
    }
    
    public boolean hasPathSum(TreeNode root, int targetSum) {
        return fun(root,targetSum);
    }
}